lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,558 @@
+}
+
+func (w *watcher) eventType(notifyType notify.Event) eventType {
+	if notifyType&w.removeEventMask() != 0 {
+		return remove
+	}
+	return nonRemove
+}
+
+func (w *watcher) VerifyConfiguration(from, to config.Configuration) error {
+	return nil
+}
+
+func (w *watcher) CommitConfiguration(from, to config.Configuration) bool {
+	found := false
+	var cfg config.FolderConfiguration
+	for _, cfg = range to.Folders {
+		if cfg.ID == w.folderID {
+			found = true
+		}
+	}
+	if !found {
+		// Nothing to do, model will soon stop this service
+		return true
+	}
+	select {
+	case w.configUpdate <- cfg:
+	case <-w.stop:
+	}
+	return true
+}
+
+func (w *watcher) updateConfig(cfg config.FolderConfiguration) {
+	w.folderPath = filepath.Clean(cfg.Path())
+	w.folderDescription = cfg.Description()
+	w.notifyDelay = time.Duration(cfg.FSWatcherDelayS) * time.Second
+	w.notifyTimeout = notifyTimeout(cfg.FSWatcherDelayS)
+	w.folderIgnorePerms = cfg.IgnorePerms
+}
+
+func (dir *eventDir) eventCount() int {
+	count := len(dir.events)
+	for _, dir := range dir.dirs {
+		count += dir.eventCount()
+	}
+	return count
+}
+
+func (dir *eventDir) childCount() int {
+	return len(dir.events) + len(dir.dirs)
+}
+
+func (dir *eventDir) removeEmptyDir(path string) {
+	delete(dir.dirs, path)
+	if dir.parentDir != nil && dir.childCount() == 0 {
+		dir.parentDir.removeEmptyDir(dir.path)
+	}
+}
+
+func watchesLimitTooLowError(folder string) error {
+	// Exchange link for own documentation when available
+	return fmt.Errorf("failed to install inotify handler for folder %s. Please increase inotify limits, see https://github.com/syncthing/syncthing-inotify#troubleshooting-for-folders-with-many-files-on-linux for more information", folder)
+}
+
+func (dir eventDir) getFirstModTime() time.Time {
+	if dir.childCount() == 0 {
+		panic("bug: getFirstModTime must not be used on empty eventDir")
+	}
+	firstModTime := time.Now()
+	for _, childDir := range dir.dirs {
+		dirTime := childDir.getFirstModTime()
+		if dirTime.Before(firstModTime) {
+			firstModTime = dirTime
+		}
+	}
+	for _, event := range dir.events {
+		if event.firstModTime.Before(firstModTime) {
+			firstModTime = event.firstModTime
+		}
+	}
+	return firstModTime
+}
+
+func (dir eventDir) getEventType() eventType {
+	if dir.childCount() == 0 {
+		panic("bug: getEventType must not be used on empty eventDir")
+	}
+	var evType eventType
+	for _, childDir := range dir.dirs {
+		evType |= childDir.getEventType()
+		if evType == mixed {
+			return mixed
+		}
+	}
+	for _, event := range dir.events {
+		evType |= event.evType
+		if evType == mixed {
+			return mixed
+		}
+	}
+	return evType
+}
+
+func (evType eventType) String() string {
+	switch {
+	case evType == nonRemove:
+		return "non-remove"
+	case evType == remove:
+		return "remove"
+	case evType == mixed:
+		return "mixed"
+	default:
+		panic("bug: Unknown event type")
+	}
+}
+
+func notifyTimeout(eventDelayS int) time.Duration {
```

Comment:
Come on, let me keep a little magic :P
I will add a lengthy comment to prevent future confusion, but essentially it just gives a weak reasoning for something that is still magic.